### PR TITLE
[FIX] prioritize cli options over yaml

### PR DESCRIPTION
**Pull request by mahon94**
### Proposed change(s)

steps to reproduce the bug:

1. run exp1 for 1M steps
2. initialize exp2 from last one with yaml["checkpoint_settings"]["initialize_from"] and continue running for eg. 200k
3. resume exp2  from  cmd and the step count continues from 1M instead of 200k

solution: checkpointSetting class now prioritize the command line arguments (resume or initialize_from for now) over conflicting yaml options. yaml option is nuke with a warning

### Types of change(s)

- [x] Bug fix
- [ ] New feature
- [ ] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/main/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ ] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/main/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/main/docs/Migrating.md) (if applicable)

### CODE REVIEW

File: ml-agents/mlagents/trainers/settings.py [hunk 1]
```diff
@@ -769,6 +769,34 @@ def maybe_init_path(self) -> Optional[str]:
     def run_logs_dir(self) -> str:
         return os.path.join(self.write_path, "run_logs")
 
+    def fix_resume_init(self) -> None:
+        """priorotize explicit command line resume/init over conflicting yaml options.
```
Comment (ervteng) [inline ml-agents/mlagents/trainers/settings.py:773]:
```suggestion
        """
        prioritize explicit command line resume/init over conflicting yaml options.
```

File: ml-agents/mlagents/trainers/settings.py [hunk 2]
```diff
@@ -769,6 +769,34 @@ def maybe_init_path(self) -> Optional[str]:
...
+        if both resume/init are set at one place use resume"""
+        _non_default_args = DetectDefault.non_default_args
+        if "resume" in _non_default_args:
+            if (
+                "initialize_from" in _non_default_args
+                or self.initialize_from is not None
+            ):
+                logger.warning(
+                    f"both 'resume' and 'initialize_from={self.initialize_from}' are set!"
+                    f" current run will be resumed ignoring initialization."
```
Comment (ervteng) [inline ml-agents/mlagents/trainers/settings.py:783]:
```suggestion
                    f" Current run will be resumed ignoring initialization."
```

File: ml-agents/mlagents/trainers/settings.py [hunk 3]
```diff
@@ -769,6 +769,34 @@ def maybe_init_path(self) -> Optional[str]:
...
+        """priorotize explicit command line resume/init over conflicting yaml options.
+        if both resume/init are set at one place use resume"""
+        _non_default_args = DetectDefault.non_default_args
+        if "resume" in _non_default_args:
+            if (
+                "initialize_from" in _non_default_args
+                or self.initialize_from is not None
+            ):
+                logger.warning(
+                    f"both 'resume' and 'initialize_from={self.initialize_from}' are set!"
```
Comment (ervteng) [inline ml-agents/mlagents/trainers/settings.py:782]:
```suggestion
                    f"Both 'resume' and 'initialize_from={self.initialize_from}' are set!"
```

File: ml-agents/mlagents/trainers/settings.py [hunk 4]
```diff
@@ -769,6 +769,34 @@ def maybe_init_path(self) -> Optional[str]:
...
+                    f" current run will be resumed ignoring initialization."
+                )
+                self.initialize_from = parser.get_default("initialize_from")
+        elif "initialize_from" in _non_default_args:
+            if self.resume is True:
```
Comment (ervteng) [inline ml-agents/mlagents/trainers/settings.py:787]:
```suggestion
            if self.resume:
```

File: ml-agents/mlagents/trainers/settings.py [hunk 5]
```diff
@@ -769,6 +769,34 @@ def maybe_init_path(self) -> Optional[str]:
...
+                logger.warning(
+                    f"both 'resume' and 'initialize_from={self.initialize_from}' are set!"
+                    f"{self.run_id} is initialized_from {self.initialize_from} and resume will be ignored."
+                )
+                self.resume = parser.get_default("resume")
+        elif self.resume is True and self.initialize_from is not None:
```
Comment (ervteng) [inline ml-agents/mlagents/trainers/settings.py:793]:
```suggestion
        else self.resume is True and self.initialize_from is not None:
```

File: ml-agents/mlagents/trainers/settings.py [hunk 6]
```diff
@@ -769,6 +769,34 @@ def maybe_init_path(self) -> Optional[str]:
     def run_logs_dir(self) -> str:
         return os.path.join(self.write_path, "run_logs")
 
+    def fix_resume_init(self) -> None:
+        """priorotize explicit command line resume/init over conflicting yaml options.
+        if both resume/init are set at one place use resume"""
+        _non_default_args = DetectDefault.non_default_args
+        if "resume" in _non_default_args:
+            if (
+                "initialize_from" in _non_default_args
```
Comment (ervteng) [inline ml-agents/mlagents/trainers/settings.py:778]: Is this condition needed? initialize_from will not be None if it is specified on the CLI.

File: ml-agents/mlagents/trainers/settings.py [hunk 7]
```diff
@@ -769,6 +769,34 @@ def maybe_init_path(self) -> Optional[str]:
     def run_logs_dir(self) -> str:
         return os.path.join(self.write_path, "run_logs")
 
+    def fix_resume_init(self) -> None:
```
Comment (ervteng) [inline ml-agents/mlagents/trainers/settings.py:772]:
```suggestion
    def prioritize_resume_init(self) -> None:
```

File: ml-agents/mlagents/trainers/settings.py (hunk 6 above)
Comment (mahon94) [reply ml-agents/mlagents/trainers/settings.py:778]: no it's not!

File: ml-agents/mlagents/trainers/tests/test_learn.py [hunk 8]
```diff
@@ -156,6 +155,12 @@ def test_commandline_args(mock_file):
     assert opt.engine_settings.no_graphics is True
     assert opt.debug is True
     assert opt.checkpoint_settings.inference is True
+    assert opt.checkpoint_settings.resume is False
+
+    # ignore init if resume set
+    full_args.append("--resume")
+    opt = parse_command_line(full_args)
+    assert opt.checkpoint_settings.initialize_from is None  # ignore init if resume set
     assert opt.checkpoint_settings.resume is True
```
Comment (ervteng) [inline ml-agents/mlagents/trainers/tests/test_learn.py:165]: Do we need a test to check if they were both set in the YAML but not the CLI? I think we can do this in the `test_settings.py`.

**Review by ervteng — APPROVED** [review]

**Review by ervteng — APPROVED** [review]
Forgot in my last approval - add this bug fix to the CHANGELOG.md file and reference the PR before merging.

**Pull request merged**